core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -0,0 +1,271 @@
+
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = topicPartitionDataForAcquirablePartitions();
+        try {
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request with an empty response.
+                shareFetchPartitionData.future().complete(Collections.emptyMap());
+                return;
+            }
+            log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
+                    topicPartitionData, shareFetchPartitionData.groupId(), shareFetchPartitionData.fetchParams());
+
+            replicaManager.fetchMessages(
+                    shareFetchPartitionData.fetchParams(),
+                    CollectionConverters.asScala(
+                            topicPartitionData.entrySet().stream().map(entry ->
+                                    new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
+                    ),
+                    QuotaFactory.UnboundedQuota$.MODULE$,
+                    responsePartitionData -> {
+                        log.trace("Data successfully retrieved by replica manager: {}", responsePartitionData);
+                        List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
+                                responsePartitionData);
+                        processFetchResponse(shareFetchPartitionData, responseData).whenComplete(
+                                (result, throwable) -> {
+                                    if (throwable != null) {
+                                        log.error("Error processing fetch response for share partitions", throwable);
+                                        shareFetchPartitionData.future().completeExceptionally(throwable);
+                                    } else {
+                                        shareFetchPartitionData.future().complete(result);
+                                    }
+                                    // Releasing the lock to move ahead with the next request in queue.
+                                    releasePartitionsLock(shareFetchPartitionData.groupId(), topicPartitionData.keySet());
+                                });
+                        return BoxedUnit.UNIT;
+                    });
+        } catch (Exception e) {
+            // Release the locks acquired for the partitions in the share fetch request in case there is an exception
+            log.error("Error processing delayed share fetch request", e);
+            shareFetchPartitionData.future().completeExceptionally(e);
+            releasePartitionsLock(shareFetchPartitionData.groupId(), topicPartitionData.keySet());
+        }
+    }
+
+    /**
+     * Try to complete the fetch operation if we can acquire records for any partition in the share fetch request.
+     */
+    @Override
+    public boolean tryComplete() {
```

Comment:
I guess we haven't added the logic to call `DelayedShareFetch.tryComplete`?

Reply:
`DelayedShareFetch.tryComplete` will be called twice in my updated code - 
1. When you add an entry to the purgatory, this function will be called internally when `delayedShareFetchPurgatory.tryCompleteElseWatch` is called [here](https://github.com/apache/kafka/pull/16969/files#diff-1e1484dbb3d93b6596214e2ec86a1fca08e2757f89eb801cebbef91f2d395406R513).
2. When you acknowledge a topic partition for which we have a pending share fetch request [here](https://github.com/apache/kafka/pull/16969/files#diff-1e1484dbb3d93b6596214e2ec86a1fca08e2757f89eb801cebbef91f2d395406R295).